host/rpc.go

```diff
@@ -0,0 +1,258 @@
+		rpc.WriteResponseErr(stream, err)
+		return
+	}
+	defer func() {
+		sh.accounts.Refund(refundAccount, budget.Remaining())
+	}()
+
+	if err := budget.Spend(settings.RPCAccountBalanceCost); err != nil {
+		log.Warnln("failed to pay for account balance RPC:", err)
+		rpc.WriteResponseErr(stream, err)
+		return
+	}
+
+	var req rhp.RPCAccountBalanceRequest
+	if err = rpc.ReadResponse(stream, &req); err != nil {
+		log.Warnln("failed to read account balance request:", err)
+		rpc.WriteResponseErr(stream, errors.New("failed to read account balance request"))
+		return
+	}
+
+	balance, err := sh.accounts.Balance(req.AccountID)
+	if err != nil {
+		log.Warnln("failed to get account balance:", err)
+		rpc.WriteResponseErr(stream, err)
+		return
+	}
+
+	resp := &rhp.RPCAccountBalanceResponse{
+		Balance: balance,
+	}
+	if err = rpc.WriteResponse(stream, resp); err != nil {
+		log.Warnln("RPC account balance:", "failed to write account balance response:", err)
+		return
+	}
+}
+
+func (sh *SessionHandler) handleRPCFundAccount(stream *mux.Stream) {
+	log := sh.log.Scope("RPCFundAccount")
+
+	var settingsID rhp.SettingsID
+	if err := rpc.ReadObject(stream, &settingsID); err != nil {
+		log.Warnln("failed to read settings UID:", err)
+		return
+	}
+
+	log.Infoln("host read settings ID")
+
+	settings, err := sh.validSettings(settingsID)
+	if err != nil {
+		log.Warnf("settings uid %s not found:", settingsID)
+		rpc.WriteResponseErr(stream, err)
+		return
+	}
+
+	log.Infoln("got settings")
+
+	budget, refundAccount, err := sh.processPayment(stream)
+	if err != nil {
+		log.Warnln("failed to process payment:", err)
+		rpc.WriteResponseErr(stream, err)
+		return
+	}
+	defer func() {
+		sh.accounts.Refund(refundAccount, budget.Remaining())
+	}()
+
+	log.Infoln("Processed payment")
+
+	if err := budget.Spend(settings.RPCFundAccountCost); err != nil {
+		log.Warnln("failed to pay for fund account RPC:", err)
+		rpc.WriteResponseErr(stream, err)
+		return
+	}
+
+	var req rhp.RPCFundAccountRequest
+	if err = rpc.ReadResponse(stream, &req); err != nil {
+		log.Warnln("failed to read fund account request:", err)
+		rpc.WriteResponseErr(stream, errors.New("failed to read fund account request"))
+		return
+	}
+
+	fundAmount := budget.Remaining()
+	balance, err := sh.accounts.Credit(req.AccountID, fundAmount)
+	if err != nil {
+		log.Warnln("failed to fund account:", err)
+		rpc.WriteResponseErr(stream, errors.New("failed to fund account"))
+		return
+	}
+	if err := budget.Spend(fundAmount); err != nil {
+		log.Warnln("failed to spend account funding:", err)
+		rpc.WriteResponseErr(stream, err)
+		return
+	}
+
+	resp := &rhp.RPCFundAccountResponse{
+		Balance: balance,
+		Receipt: rhp.Receipt{
+			Host:      sh.privkey.PublicKey(),
+			Account:   req.AccountID,
+			Amount:    fundAmount,
+			Timestamp: time.Now(),
+		},
+	}
+
+	h := types.NewHasher()
+	resp.Receipt.EncodeTo(h.E)
+	resp.Signature = sh.privkey.SignHash(h.Sum())
```

Comment:
Probably worth defining a `hashObject(obj types.EncoderTo)` helper function. It'll be a bit less efficient, but improved readability makes it a net win imo